### PR TITLE
Authentication Enhancements

### PR DESCRIPTION
**Pull request by Nitish145**
#### Describe the changes you have made in this PR -
1. Add the following endpoint
      - `/api/v1/password/forgot` to send reset password link.
      - `/api/v1/public_key.pem` to GET jwt public key

2. Fix rubocop linting issues for authentication and users specs.

3. Add specs for the above two endpoints

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by Nitish145** [comment]
@tachyons please check!

File: app/controllers/api/v1/authentication_controller.rb [hunk 1]
```diff
@@ -29,6 +27,20 @@ def signup
...
+    @user = User.find_by!(email: params[:email])
+    # sends reset password instructions to the user's mail if exists
+    @user.send_reset_password_instructions
+    render json: { message: "password reset instructions sent to #{@user.email}" }
+  end
+
+  # GET /public_key.pem
+  def public_key
+    public_key = File.open(Rails.root.join("config", "public.pem"), "r:UTF-8")
+    send_file public_key
```
Comment (tachyons) [inline app/controllers/api/v1/authentication_controller.rb:41]: What about making this a json so that this will be easier to consume at client level

Comment (Nitish145) [reply app/controllers/api/v1/authentication_controller.rb:41]: Yes!! Would be easier..

**Comment by Nitish145** [comment]
@tachyons please check!!

File: app/controllers/api/v1/authentication_controller.rb [hunk 2]
```diff
@@ -29,6 +27,44 @@ def signup
     end
   end
 
+  # POST api/v1/oauth/login
+  def oauth_login
+    @user = User.find_by!(email: params["email"])
+    token = JsonWebToken.encode(
+      user_id: @user.id, username: @user.name, email: @user.email
+    )
+    render json: { token: token }, status: :accepted
```
Comment (tachyons) [inline app/controllers/api/v1/authentication_controller.rb:36]: Wait, you can pass arbitrary email addresses and this endpoint will return a valid token for that user !!. Big security flow

Comment (Nitish145) [reply app/controllers/api/v1/authentication_controller.rb:36]: @tachyons is that so? what we are doing here is just registering or finding the user with that email/info in the DB.. & imo devise handles email validation on its own, so it won't allow any email to generate a valid token..
Correct me if i am wrong..

Comment (Nitish145) [reply app/controllers/api/v1/authentication_controller.rb:36]: @tachyons

Comment (tachyons) [reply app/controllers/api/v1/authentication_controller.rb:36]: Where are you validating that ?

Comment (Nitish145) [reply app/controllers/api/v1/authentication_controller.rb:38]: @tachyons pushed the appropriate changes discussed, can you please once check.. if given a green signal, will add tests for the same..

File: app/controllers/api/v1/authentication_controller.rb [hunk 3]
```diff
@@ -29,27 +31,33 @@ def signup
...
-      @user = User.new(oauth_signup_params)
-      @user.password = Devise.friendly_token[0, 20]
-      @user.save!
+      @user = User.create!(
+        name: @response["name"],
+        email: @response["email"],
+        password: Devise.friendly_token[0, 20],
+        provider: params[:provider],
+        uid: @response["id"] || @response["sub"]
+      )
```
Comment (tachyons) [inline app/controllers/api/v1/authentication_controller.rb:53]: There can be scenarios like empty email from the response, for instance you can create a Facebook account with email linked to it. So handle failing to create user record with proper error message

Comment (Nitish145) [reply app/controllers/api/v1/authentication_controller.rb]: @tachyons can you please check once!!

File: app/models/user.rb [hunk 4]
```diff
@@ -67,6 +67,16 @@ def self.from_omniauth(access_token)
     user
   end
 
+  def self.from_oauth(oauth_user, provider)
+    User.create!(
+      name: oauth_user["name"],
+      email: oauth_user["email"],
```
Comment (tachyons) [inline app/models/user.rb:73]: what if email is blank ?

Comment (Nitish145) [reply app/models/user.rb:73]: @tachyons `create!` will raise validation error which would be handled in the rescue block.. with appropriate error message..

```
rescue ActiveRecord::RecordInvalid
   api_error(status: 422, errors: @user.errors)
end
```

Comment (tachyons) [reply app/models/user.rb:73]: @user is not assigned in this case, write a spec for the scenario where email is blank

Comment (Nitish145) [reply app/models/user.rb:73]: Oh is it.. will cover that case..

Comment (Nitish145) [reply app/models/user.rb:73]: Thanks @tachyons for this insight!! learned something new about how `bang!` only raises an exception.. Thanks :v:

Comment (Nitish145) [reply app/models/user.rb:73]: btw made the required changes and added specs.. please check @tachyons

File: spec/requests/api/v1/authentication_controller/oauth_signup_spec.rb [hunk 5]
```diff
@@ -45,6 +45,20 @@
...
+      before do
+        post "/api/v1/oauth/signup", params: {
+          "access_token": "empty_email_token",
+          "provider": "google"
+        }, as: :json
+      end
+
+      it "return status 422 and should have jsonapi errors" do
+        expect(response).to have_http_status(422)
+        expect(response.parsed_body).to have_jsonapi_errors
```
Comment (tachyons) [inline spec/requests/api/v1/authentication_controller/oauth_signup_spec.rb:58]: Can you check the exact error message, need to make end user aware that signup failed due to missing email in the linked account

Comment (Nitish145) [reply spec/requests/api/v1/authentication_controller/oauth_signup_spec.rb:58]: Done! @tachyons

**Comment by Nitish145** [comment]
@tachyons anymore changes here ?

**Pull request merged**